Add reader debug string as exception context reader

oerling opened the pull request:
Add reader debug string as exception context  

Adds an ExceptionContext around DataSource::next. Records the
exception context string in SelectiveStructColumnReader so that an
eventual LazyVector load can find this string and show it as context
of an eventual exception. Allows localizing the file and Task id
related to the error.

Yuhta commented:
Is the xsimd upgrade intended?  We may want to keep the version synchronized between this submodule and the one in `fbsource/third-party`

kagamiori commented on velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:
```diff
@@ -61,6 +61,8 @@ SelectiveStructColumnReader::SelectiveStructColumnReader(
     childSpec->setSubscript(children_.size() - 1);
   }
+  thread_local std::string readerDebugString;
+  debugString_ = readerDebugString;
```
Why do we need debugString_ instead of just using readerDebugString? Also, it looks like we don't initialize readerDebugString? (Maybe I didn't understand thread_local. Could you explain a bit about this?)

kagamiori reviewed (commented):
Hi Orri, Thank you for working on this! I have a question about readerDebugString.

oerling commented:
Updated the PR. Files were a different version. Maybe now it makessense.

kagamiori commented on velox/exec/TableScan.cpp:
```diff
@@ -89,12 +89,26 @@ RowVectorPtr TableScan::getOutput() {
             "Got splits with different connector IDs");
       }
 
+      thread_local std::string readerDebugString = fmt::format(
```
Can readerDebugString be removed? It looks the same as debugString_.

kagamiori commented on velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp:
```diff
@@ -33,7 +33,8 @@ SelectiveStructColumnReader::SelectiveStructColumnReader(
           std::move(flatMapContext)),
-      requestedType_{requestedType} {
+      requestedType_{requestedType},
+      debugString_(getExceptionContext().message()) {
```
Why does SelectiveStructColumnReader get its debugString_ from the current ExceptionContext? Is this ExceptionContext supposed to be the one set in TableScan::getOutput()? Might there be cases where the current ExceptionContext is not set by TableScan as intended?

kagamiori reviewed (commented):
Hi Orri, Thank you for updating the code. I'm trying to understand the code and have a few questions. Also, could you add a unit test if possible?

oerling commented:
Updated. Removed the duplicate debug string . 

 

The reader tree is made on first use inside DataSource::next. TableScan puts the string in the context around that call. Lazy vector loaders get the context string from the struct reader whenever they are called, which is not inside DataSource::next. Reader errors with filtered columns etc take place inside DataSource::next so see the string from the context.

 

If there is no context, as for example in tests,  the string just ends up empty.

 

Places where reader trees are inited from the query should set the context. Right now there is only this one.

 

 

 

 

From: kagamiori ***@***.***> 
Sent: Thursday, May 5, 2022 9:23 PM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Add reader debug string as exception context reader (PR #1522)

 

@kagamiori commented on this pull request.

Hi Orri, Thank you for updating the code. I'm trying to understand the code and have a few questions. Also, could you add a unit test if possible?

  _____  

In velox/exec/TableScan.cpp <https://github.com/facebookincubator/velox/pull/1522#discussion_r866478054> :

> @@ -89,12 +89,26 @@ RowVectorPtr TableScan::getOutput() {
             "Got splits with different connector IDs");
       }
 
+      thread_local std::string readerDebugString = fmt::format(

Can readerDebugString be removed? It looks the same as debugString_.

  _____  

In velox/dwio/dwrf/reader/SelectiveStructColumnReader.cpp <https://github.com/facebookincubator/velox/pull/1522#discussion_r866478986> :

> @@ -33,7 +33,8 @@ SelectiveStructColumnReader::SelectiveStructColumnReader(
           scanSpec,
           dataType->type,
           std::move(flatMapContext)),
-      requestedType_{requestedType} {
+      requestedType_{requestedType},
+      debugString_(getExceptionContext().message()) {

Why does SelectiveStructColumnReader get its debugString_ from the current ExceptionContext? Is this ExceptionContext supposed to be the one set in TableScan::getOutput()? Might there be cases where the current ExceptionContext is not set by TableScan as intended?

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1522#pullrequestreview-964164424> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPT5ZJBYHQ7MYAZZTXPTVISNAXANCNFSM5U42BNJQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

kagamiori reviewed (approved):
Thank you for the explanation! Looks good to me. By the way, why LazyVector load is not inside DataSource::next but other reader errors are in DataSource::next? Is it because it's "lazy"? If I'd like to learn more about lazy vectors, which files do you recommend to read?

oerling commented:
The load is not inside DataSource::next because it is on first use wherever. You can give me a call this PM and I can explain this. Basically LazyVector.h and references to loadedVector, loadedChild, ensureFieldLoaded all around.

 

Thanks

Orri

 

 

From: kagamiori ***@***.***> 
Sent: Friday, May 6, 2022 4:25 PM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Add reader debug string as exception context reader (PR #1522)

 

@kagamiori approved this pull request.

Thank you for the explanation! Looks good to me. By the way, why LazyVector load is not inside DataSource::next but other reader errors are in DataSource::next? Is it because it's "lazy"? If I'd like to learn more about lazy vectors, which files do you recommend to read?

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1522#pullrequestreview-965251526> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPTYGFB4TEV6FSIZLKTTVIWS43ANCNFSM5U42BNJQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

oerling commented:
The mentioned lines  are just left over from a previous attempt at doing the same thing. Deleted.
